[#12467] Add Sort Functions to Admin Sessions Page

singhabhyudita opened the pull request:
<!-- Before opening a PR, please ensure you have read our contributor guidelines -->
<!-- at https://teammates.github.io/teammates/process.html#step-4-submit-a-pr. -->

<!-- PR title: Copy-and-paste the name of the issue this PR is fixing, -->
<!-- and include the issue number in front in square brackets. -->
<!-- e.g. [#3942] Remove unnecessary System.out.printlns from Java files -->

<!-- Add the issue number to the "Fixes" keyword below. -->
Fixes #12467 

**Outline of Solution**

I tried using the `sortable-table` component. I have created a method `populateSortableTable()` which is called in the `getFeedbackSessions()` method. There are a few formatting changes that had to be done.  The creator email Id is no longer a hyperlink and same goes for response rate. Previously the response rate was displayed by clicking on the show button, but due to the restrictions of the `sortable-table` component, I am unable to do so hence I have left it empty. 

Any suggestions regarding how to display the response rate are welcome. 

Also for the other two functions, one can be sort by institute name but I am not too sure about the other one. 

<img width="754" alt="Screenshot 2023-07-13 at 1 19 14 AM" src="https://github.com/TEAMMATES/teammates/assets/60813273/a70b4368-99c5-4345-a3f4-1940bbc97a78">





<!-- Give a brief description of how you solved the issue. -->
<!-- If the solution includes any changes in UI, do also attach screenshots of the new UI. -->

singhabhyudita commented:
> Hi @singhabhyudita, thank you for your interest in contributing to TEAMMATES! However, your PR does not appear to follow our [contribution guidelines](https://teammates.github.io/teammates/process.html#step-4-submit-a-pr):
> 
> * Description must reference the issue number the PR is fixing, e.g. `Fixes #<issue-number>` (or `Part of #<issue-number>` if the PR does not address the issue fully)
> 
> Please address the above before we proceed to review your PR.

Fixed

singhabhyudita commented:
Sure I will take a look. Also there is one very trivial issue here, the sortable-table component is not working. I am trying to sort the tables by start / end time but it is unresponsive.

Zxun2 commented on src/web/app/pages-admin/admin-sessions-page/admin-sessions-page.component.ts:
```diff
@@ -85,6 +101,31 @@ export class AdminSessionsPageComponent implements OnInit {
+            { displayValue: session.ongoingSession.sessionStatus },
+            { displayValue: '['+session.ongoingSession.courseId+'] '+session.ongoingSession.feedbackSessionName },
+            { displayValue: ''},
+            { displayValue: session.startTimeString },
+            { displayValue: session.endTimeString },
```
I think the issue lies here! You need to fill in the `value` field which is used for sorting.

singhabhyudita replied:
You're right, thanks!

singhabhyudita commented:
Current changes:

<img width="1440" alt="Screenshot 2023-07-15 at 6 02 10 PM" src="https://github.com/TEAMMATES/teammates/assets/60813273/fe11661d-e12c-4bca-95bb-7d09ca7fda16">

weiquu commented:
Hi @singhabhyudita, do note that you can run `npm run lint` to view the errors (or click on "Details" next to the failing checks). The E2E test is also currently failing due to the table lacking the correct `id` attribute.

singhabhyudita commented:
> Hi @singhabhyudita, do note that you can run `npm run lint` to view the errors (or click on "Details" next to the failing checks). The E2E test is also currently failing due to the table lacking the correct `id` attribute.

Hi, could you tell me more about the `id` attribute. Where is it needed?

weiquu commented:
> Hi, could you tell me more about the `id` attribute. Where is it needed?

@singhabhyudita you have it in the right place, but it's misspelled! Should be "ong**o**ing", not "onging" (:

singhabhyudita commented:
I think the e2e tests are failing because in [AdminSessionsPageE2ETest.java](https://github.com/TEAMMATES/teammates/blob/2fb5035ca46602be6714bfce2b604552ba827408/src/e2e/java/teammates/e2e/cases/AdminSessionsPageE2ETest.java), the expected output format for the `openSessionCells` has 

```
  String[] openSessionCells = {
                "[Opened]",
                String.format("[%s] %s", openFeedbackSession.getCourseId(),
                        openFeedbackSession.getFeedbackSessionName()),
                "Show",
                formatDateTime(instant3DaysAgo, tableTimezone),
                formatDateTime(instant3DaysLater, tableTimezone),
                openFeedbackSession.getCreatorEmail(),
        };
```
whereas currently I am directly displaying the `responseRate` as compared to the previous Show button. 
<img width="134" alt="Screenshot 2023-07-17 at 8 54 03 AM" src="https://github.com/TEAMMATES/teammates/assets/60813273/427a7cde-314d-478b-9761-26190083f2cd">

I tried to change the test file but the `FeedbackSessionAttributes` class doesn't have a `getResponseRate()` method. 
To change that we would need to add a responseRate attribute to both the `FeedbackSession` entity and `FeedbackSessionAttributes`. 

Or if we can somehow figure out how to add a button in the `sortable-table` component, then that would work too.

weiquu commented:
Hi @singhabhyudita, unfortunately I don't think we should be getting the response rates for all sessions when we populate the table... it will likely cause a huge performance issue, especially given that there might be many sessions ongoing during the range selected.

@Zxun2 has kept the original response rate behaviour by introducing a `cell-with-response-rate` component (`tm-response-rate`) and using that in the `sessions-table` component. When setting the row data of the table,  he calls the `createCellWithResponseRateComponent` (relevant lines: 254 and 293-311) which then creates the `tm-response-rate` component with the original functionality. Note that upon clicking the "Show" button, it triggers the `loadResponseRateEvent` which in turn calls the `loadResponseRateEventHandler` with the relevant index. (@Zxun2 lmk if I got this wrong!)

I think you can try doing something similar here - no need to worry about code duplication for now. Alternatively, you can wait until #12501 gets merged, which I suppose might be about next week since it's a large PR (will get to reviewing it this week).

singhabhyudita commented:
Hi @weiquu @Zxun2 is there any way to specify `@Output` values in the customComponent's componentData?
Passing the `getResponseRate()` method as an input into the custom component does not work. It shows a `TypeError: Cannot read property of undefined` for the `loadSessionStatistics`.  So I shifted the method into the `ResponseRateComponent`.

I am also currently trying to work on another issue, Refactoring of sortable tables for the instructor courses page. Even there I tried to pass the `onDelete()` method into the custom component and it is showing the same error. Here shifting the method into another component is not possible so I was wondering if we can use the event emitters to call these methods.

singhabhyudita commented:
Pinging @weiquu @domlimm for review!

weiquu commented:
Hi @singhabhyudita, a couple of comments to work on first:

Initially, the "Institution Name" button is disabled, indicating that the panels are sorted by institution name. However, that's not the case - clicking on "Institution Sessions Total" and back to "Institution Name" gives in a different sort order. Do ensure that the initial sort order is correct.

The sorting by start and end dates are also incorrect. Intuitively, the sort should be by a date comparison, so for example, ascending means that the dates should get later further down the rows. However, it seems that the current comparison is just based on strings, meaning that (for an ascending sort) Fridays will come first, then Mondays, etc, regardless of the actual date.

Will take a closer look at the code once these are fixed

singhabhyudita commented:
> Hi @singhabhyudita, a couple of comments to work on first:
> 
> Initially, the "Institution Name" button is disabled, indicating that the panels are sorted by institution name. However, that's not the case - clicking on "Institution Sessions Total" and back to "Institution Name" gives in a different sort order. Do ensure that the initial sort order is correct.
> 
> The sorting by start and end dates are also incorrect. Intuitively, the sort should be by a date comparison, so for example, ascending means that the dates should get later further down the rows. However, it seems that the current comparison is just based on strings, meaning that (for an ascending sort) Fridays will come first, then Mondays, etc, regardless of the actual date.
> 
> Will take a closer look at the code once these are fixed

For the start and end dates I just used the existing enums. On taking a closer look I found out that the `SESSION_START_DATE` and `SESSION_END_DATE` are being sorted lexicographically in the table-comparator service file.

weiquu commented:
> For the start and end dates I just used the existing enums. On taking a closer look I found out that the `SESSION_START_DATE` and `SESSION_END_DATE` are being sorted lexicographically in the table-comparator service file.

You're right that they are being sorted lexicographically; however, the underlying data is being stored as a number (timestamp) rather than the string. Hence, when sorted, the numbers are compared which produces the correct result. You can see this in action by going to the `compareLexicographically` method of the `table-comparator.service.ts` and logging `strA` and `strB`. When the admin sessions table is sorted, they output a date string; when say the sessions table in the instructor home page is sorted, they output a number representing the date.

My suggestion is to take a closer look at how the sessions table in the instructor home page is handled, including the `createDateCellWithToolTip` which you'll see transforms the timestamp (a number) into the string value to be displayed to the user. Some changes will probably be needed to be made to the `OngoingSessionsModel` (which I'm not sure why is defined in the response rate cell component... would have thought leaving it in the admin sessions page component would be better). I think now only the string is stored - either store both, or only store the `endTime`/`startTime` and then convert it to a string later on.

singhabhyudita commented:
> > For the start and end dates I just used the existing enums. On taking a closer look I found out that the `SESSION_START_DATE` and `SESSION_END_DATE` are being sorted lexicographically in the table-comparator service file.
> 
> You're right that they are being sorted lexicographically; however, the underlying data is being stored as a number (timestamp) rather than the string. Hence, when sorted, the numbers are compared which produces the correct result. You can see this in action by going to the `compareLexicographically` method of the `table-comparator.service.ts` and logging `strA` and `strB`. When the admin sessions table is sorted, they output a date string; when say the sessions table in the instructor home page is sorted, they output a number representing the date.
> 
> My suggestion is to take a closer look at how the sessions table in the instructor home page is handled, including the `createDateCellWithToolTip` which you'll see transforms the timestamp (a number) into the string value to be displayed to the user. Some changes will probably be needed to be made to the `OngoingSessionsModel` (which I'm not sure why is defined in the response rate cell component... would have thought leaving it in the admin sessions page component would be better). I think now only the string is stored - either store both, or only store the `endTime`/`startTime` and then convert it to a string later on.

Can we maybe create a function that converts the dates into a `Date` object and then sorts it?

I will also look at the instructor home page sessions table. 
Also the `OngoingSessionsModel` has been shifted to `cellWithResponseRateComponent` because there was a dependency cycle being created since the ongoing sessions object was being passed into the response rate component.

weiquu commented:
> Can we maybe create a function that converts the dates into a `Date` object and then sorts it?

Don't see any issue with that, though I'm not sure if it's easier or cleaner than the current approach. (Do note that, although unlikely, there might be an issue with this that might be uncovered later.)

> Also the `OngoingSessionsModel` has been shifted to `cellWithResponseRateComponent` because there was a dependency cycle being created since the ongoing sessions object was being passed into the response rate component.

Hmm haven't taken a close look at the code, but is there any difference with the `cell-with-response-rate.component.ts` being used in the `sessions-table` component? For quick reference, here are the inputs being used there:
```
  @Input() responseRate: string = '';
  @Input() idx: number = 0;
  @Input() empty: boolean = false;
  @Input() isLoading: boolean = false;
  @Input() onClick: () => void = () => { };
```
My intuition is that the implementation of this should be the same, not too sure why the ongoing sessions object is being passed in. Do let me know if I've missed something (and feel free to modify or add to the fields in the ongoing sessions object).

singhabhyudita commented:
The implementation here is different because the in the other case the changes are being made to the sessions table component (which is using the sortable-table component) , which is in turn being used in the instructor sessions page. In this case the admin sessions page is directly using the sortable-table component. Probably that's why createCellWithResponseRateComponent has a different implementations. 


I have fixed the sorting issue by converting the date into milliseconds. Also fixed the initial sort issue.

singhabhyudita commented:
Pinging @weiquu @domlimm  for review

singhabhyudita commented:
> Pinging @weiquu @domlimm for review

@weiquu  @domlimm  @zhaojj2209

weiquu commented:
Hi @singhabhyudita, apologies for the delay - been rather busy. Thanks for the changes to the sorting by start and end date within tables, those work fine now. Before we take a closer look at the actual implementation, could you help to fix the following issues?

1. Sort button does not reset

I first sort the panels by "Institution Sessions Total", which gives the following result:
<img width="1278" alt="1 - sort by sessions total" src="https://github.com/TEAMMATES/teammates/assets/48304907/36d0dad8-128b-470a-bf95-765318dd832c">

Afterwards, I increase the date range and click "Filter by Range". This gives the following result:
<img width="1262" alt="1 - incorrect sort" src="https://github.com/TEAMMATES/teammates/assets/48304907/5517d2c3-4009-443e-8ebd-8708898de8d9">

As can be seen, even though the "Institution Sessions Total" button is still shown to be selected, the panels are not sorted accordingly. I think that either the Sort By should be sorted by whatever is currently selected, or it should reset to "Institution Name" (and the button should also reset). Both approaches are fine.

Further note that the panels are also not correctly sorted by Institution Name ("test" should be after "TEAMMATES Test Institute 1"). This is likely related to the next issue:

2. Issue with initial sort by institution name

After setting a date range and pressing "Filter by Range", I get the following sort:
<img width="1288" alt="2 - initial" src="https://github.com/TEAMMATES/teammates/assets/48304907/284c71aa-e1bd-44bb-bef1-6d0369e0a981">

However, the sort above is incorrect. After pressing "Institution Sessions Total" and then back to "Institution Name", I get the following correct sort:
<img width="1284" alt="2 - correctly sorted" src="https://github.com/TEAMMATES/teammates/assets/48304907/83d885df-4304-44a4-8c94-be273c9b2e91">

The sort should be correct even upon the initial loading.

3. Panels not closing properly

I first set the date range and press "Filter by Range". Then, I sort by "Institution Sessions Total". Afterwards, I try to close the first panel; however, the panel does not close, even though the icon shows that it is closed. Upon pressing it again (to open the already open panel), I get what appears to be 2 tables under 1 panel:
<img width="1298" alt="3 - 2 tables" src="https://github.com/TEAMMATES/teammates/assets/48304907/41a7e430-5980-40a4-8583-28a4fbf8cf62">

Furthermore, after trying to close the panel, one of the tables disappears, while one remains:
<img width="1259" alt="3 - icon shows closed but open" src="https://github.com/TEAMMATES/teammates/assets/48304907/b98265a4-e070-4315-8ea7-68bf753e4f6e">

Note that the icon shows that the panel is closed, yet there is a table still visible beneath it.

Do let me know if you have any trouble recreating these issues.

cedricongjh commented:
hey @singhabhyudita, we will be closing this PR for now, if you're working on it still, do re-open it!

singhabhyudita commented:
Hi @cedricongjh , have been occupied for a while. Will start working on it again by the end of this week. Thanks.

singhabhyudita commented:
Hi @weiquu @cedricongjh can we reopen this PR. I have started working on it again.

singhabhyudita commented:
> Hi @singhabhyudita, apologies for the delay - been rather busy. Thanks for the changes to the sorting by start and end date within tables, those work fine now. Before we take a closer look at the actual implementation, could you help to fix the following issues?
> 
> 1. Sort button does not reset
> 
> I first sort the panels by "Institution Sessions Total", which gives the following result: <img alt="1 - sort by sessions total" width="1278" src="https://user-images.githubusercontent.com/48304907/259812219-36d0dad8-128b-470a-bf95-765318dd832c.png">
> 
> Afterwards, I increase the date range and click "Filter by Range". This gives the following result: <img alt="1 - incorrect sort" width="1262" src="https://user-images.githubusercontent.com/48304907/259812312-5517d2c3-4009-443e-8ebd-8708898de8d9.png">
> 
> As can be seen, even though the "Institution Sessions Total" button is still shown to be selected, the panels are not sorted accordingly. I think that either the Sort By should be sorted by whatever is currently selected, or it should reset to "Institution Name" (and the button should also reset). Both approaches are fine.
> 
> Further note that the panels are also not correctly sorted by Institution Name ("test" should be after "TEAMMATES Test Institute 1"). This is likely related to the next issue:
> 
> 2. Issue with initial sort by institution name
> 
> After setting a date range and pressing "Filter by Range", I get the following sort: <img alt="2 - initial" width="1288" src="https://user-images.githubusercontent.com/48304907/259813117-284c71aa-e1bd-44bb-bef1-6d0369e0a981.png">
> 
> However, the sort above is incorrect. After pressing "Institution Sessions Total" and then back to "Institution Name", I get the following correct sort: <img alt="2 - correctly sorted" width="1284" src="https://user-images.githubusercontent.com/48304907/259813262-83d885df-4304-44a4-8c94-be273c9b2e91.png">
> 
> The sort should be correct even upon the initial loading.
> 
> 3. Panels not closing properly
> 
> I first set the date range and press "Filter by Range". Then, I sort by "Institution Sessions Total". Afterwards, I try to close the first panel; however, the panel does not close, even though the icon shows that it is closed. Upon pressing it again (to open the already open panel), I get what appears to be 2 tables under 1 panel: <img alt="3 - 2 tables" width="1298" src="https://user-images.githubusercontent.com/48304907/259813872-41a7e430-5980-40a4-8583-28a4fbf8cf62.png">
> 
> Furthermore, after trying to close the panel, one of the tables disappears, while one remains: <img alt="3 - icon shows closed but open" width="1259" src="https://user-images.githubusercontent.com/48304907/259814351-b98265a4-e070-4315-8ea7-68bf753e4f6e.png">
> 
> Note that the icon shows that the panel is closed, yet there is a table still visible beneath it.
> 
> Do let me know if you have any trouble recreating these issues.

@weiquu thank you for pointing this out. I have fixed the first 2 issues. The panel issue however is something weird. When I try to close the first few tables the issue exists but now for the last table. When I change the sorting option while the tables are expanded and then try to close the tables the issue exists. But when the table is collapsed and I switch the option it works fine again. I am not able to understand where the problem lies exactly because it is not consistent among all tables.
